### PR TITLE
Modular backend - LoRA/LyCORIS

### PR DESCRIPTION
**Pull request by StAlKeR7779**
## Summary

Code for lora patching from #6577.
Additionally made it the way, that lora can patch not only `weight`, but also `bias`, because saw some loras which doing it.

## Related Issues / Discussions

#6606 
https://invokeai.notion.site/Modular-Stable-Diffusion-Backend-Design-Document-e8952daab5d5472faecdc4a72d377b0d

## QA Instructions

Run with and without set `USE_MODULAR_DENOISE` environment.

## Merge Plan

Replace old lora patcher with new after review done.
If you think that there should be some kind of tests - feel free to add.

## Checklist

- [x] _The PR has a short but descriptive title, suitable for a changelog_
- [ ] _Tests added / updated (if applicable)_
- [ ] _Documentation added / updated (if applicable)_

### CODE REVIEW

File: invokeai/backend/stable_diffusion/extensions/lora_patcher.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        yield self._patch_model(
+            model=unet,
+            prefix=self._prefix,
+            loras=_lora_loader(),
+            cached_weights=cached_weights,
+        )
+
+    @classmethod
+    @contextmanager
+    def static_patch_model(
```
Comment (RyanJDick) [inline invokeai/backend/stable_diffusion/extensions/lora_patcher.py:49]: Unused?

**Review by RyanJDick — CHANGES_REQUESTED** [review]

File: invokeai/backend/lora.py [hunk 2]
```diff
@@ -186,37 +213,41 @@ def __init__(
...
-            self.rank = values["lokr_w2_b"].shape[0]
+        if self.w1_b is not None:
+            self.rank = self.w1_b.shape[0]
+        elif self.w2_b is not None:
+            self.rank = self.w2_b.shape[0]
         else:
             self.rank = None  # unscaled
 
-    def get_weight(self, orig_weight: Optional[torch.Tensor]) -> torch.Tensor:
+        # Although lokr_t1 not used in algo, it still defined in LoKR weights
```
[the remark this answers is not in the document]
Comment (StAlKeR7779) [reply invokeai/backend/lora.py:235]: I can reference to code, but it really as it said, there no reference anywhere in LoKR LyCORIS code to `lokr_t1`, but it presented in weights list:
https://github.com/KohakuBlueleaf/LyCORIS/blob/main/lycoris/modules/lokr.py
```py
    weight_list = [
        "lokr_w1",
        "lokr_w1_a",
        "lokr_w1_b",
        "lokr_w2",
        "lokr_w2_a",
        "lokr_w2_b",
        "lokr_t1",
        "lokr_t2",
        "alpha",
        "dora_scale",
    ]
```

File: invokeai/backend/lora.py [hunk 3]
```diff
@@ -60,6 +70,15 @@ def to(self, device: Optional[torch.device] = None, dtype: Optional[torch.dtype]
...
             self.bias = self.bias.to(device=device, dtype=dtype)
 
+    def check_keys(self, values: Dict[str, torch.Tensor], known_keys: Set[str]):
+        all_known_keys = known_keys | {"alpha", "bias_indices", "bias_values", "bias_size"}
+        unknown_keys = set(values.keys()) - all_known_keys
+        if unknown_keys:
+            # TODO: how to warn log?
+            print(
+                f"[WARN] Unexpected keys found in LoRA/LyCORIS layer, model might work incorrectly! Keys: {unknown_keys}"
+            )
```
[the remark this answers is not in the document]
Comment (StAlKeR7779) [reply invokeai/backend/lora.py:80]: Yeah I wanted to do it, but don't know how(except for through InvocationContext), maybe you can hint?

File: invokeai/backend/lora.py [hunk 4]
```diff
@@ -46,9 +46,19 @@ def __init__(
...
 
+    def get_bias(self, orig_bias: torch.Tensor) -> Optional[torch.Tensor]:
+        return self.bias
+
+    def get_parameters(self, orig_module: torch.nn.Module) -> Dict[str, torch.Tensor]:
+        params = {"weight": self.get_weight(orig_module.weight)}
+        bias = self.get_bias(orig_module.bias)
+        if bias is not None:
+            params["bias"] = bias
+        return params
```
[the remark this answers is not in the document]
Comment (StAlKeR7779) [reply invokeai/backend/lora.py:60]: In current code I added bias support only to full/diff layer, as it layer already exists in code.
I not saw any such loras, just done some kind of "sync" with LyCORIS code where said that bias can be used:
https://github.com/KohakuBlueleaf/LyCORIS/blob/main/lycoris/modules/full.py#L105

But I know one lora file where used patch for normalization modules and which contains bias, this layer type not supported currently in Invoke:
https://github.com/KohakuBlueleaf/LyCORIS/blob/main/lycoris/modules/norms.py

File: invokeai/backend/lora.py [hunk 5]
```diff
@@ -459,23 +490,23 @@ def from_checkpoint(
...
+            elif "lokr_w1" in values or "lokr_w1_a" in values:
                 layer = LoKRLayer(layer_key, values)
 
             # diff
             elif "diff" in values:
                 layer = FullLayer(layer_key, values)
 
             # ia3
-            elif "weight" in values and "on_input" in values:
+            elif "on_input" in values:
```
[the remark this answers is not in the document]
Comment (StAlKeR7779) [reply invokeai/backend/lora.py:509]: As I said in commit title, it's just sync with logic in LyCORIS repository, you can open different format layers here:
https://github.com/KohakuBlueleaf/LyCORIS/tree/main/lycoris/modules
And see which keys used for detection of each format, for example LoKR:
https://github.com/KohakuBlueleaf/LyCORIS/blob/main/lycoris/modules/lokr.py#L46
```py
weight_list_det = ["lokr_w1", "lokr_w1_a"]
```

I thought that such sync ok as they defining this formats.

File: invokeai/backend/stable_diffusion/extensions/lora_patcher.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+        finally:
+            with torch.no_grad():
+                for param_key in modified_cached_weights:
+                    model.get_parameter(param_key).copy_(cached_weights[param_key])
+                for param_key, weight in modified_weights.items():
+                    model.get_parameter(param_key).copy_(weight)
+
+    @classmethod
+    def _patch_model(
```
[the remark this answers is not in the document]
Comment (StAlKeR7779) [reply invokeai/backend/stable_diffusion/extensions/lora_patcher.py:73]: Yes, as I mentioned in merge plan - I planning to replace calls to ModelPatcher with calls to `static_patch_model` in this class.
For now not changed, so that we can easily compare that results are same.

File: invokeai/backend/stable_diffusion/extensions/lora_patcher.py (hunk 1 above)
Comment (StAlKeR7779) [reply invokeai/backend/stable_diffusion/extensions/lora_patcher.py:49]: See comment about `_patch_model`.
Shortly - will be used by compel nodes when we test that output fully same(I locally tested, just want to someone else test too)

File: invokeai/backend/stable_diffusion/extensions/lora_patcher.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+    @contextmanager
+    def patch_unet(self, unet: UNet2DConditionModel, cached_weights: Optional[Dict[str, torch.Tensor]] = None):
+        def _lora_loader() -> Iterator[Tuple[LoRAModelRaw, float]]:
+            for lora in self._loras:
+                lora_info = self._node_context.models.load(lora.lora)
+                lora_model = lora_info.model
+                yield (lora_model, lora.weight)
+                del lora_info
+            return
```
[the remark this answers is not in the document]
Comment (StAlKeR7779) [reply invokeai/backend/stable_diffusion/extensions/lora_patcher.py:38]: About method - ok, it could be done as separate method, I agree.
About separate lora extensions, now as you mentioned it looks reasonable. Current implementation was made before realizing how to handle IC-Light too.
So this all collected in one patcher to collect all changed weights in one place, but now `ExtensionsManager` collects weights and each lora can become separate extension.

File: invokeai/backend/stable_diffusion/extensions/base.py [hunk 8]
```diff
@@ -56,5 +56,17 @@ def patch_extension(self, ctx: DenoiseContext):
         yield None
 
     @contextmanager
-    def patch_unet(self, unet: UNet2DConditionModel, cached_weights: Optional[Dict[str, torch.Tensor]] = None):
-        yield None
+    def patch_unet(
+        self, unet: UNet2DConditionModel, cached_weights: Optional[Dict[str, torch.Tensor]] = None
+    ) -> Tuple[Set[str], Dict[str, torch.Tensor]]:
```
[the remark this answers is not in the document]
Comment (StAlKeR7779) [reply invokeai/backend/stable_diffusion/extensions/base.py:61]: Implemented another idea, wait for new comments.

**Review by RyanJDick — APPROVED** [review]
Looks good!

I tested the following test cases on both _old_invoke and _new_invoke against main:
- [x] LoRA patching and unpatching speeds are the same
- [x] Images generated with LoRA are unchanged
- [x] Images generated after removing a LoRA are unchanged (i.e. the LoRA is being unpatched properly).
- [x] Can stack multiple LoRAs

I left one minor comment. I'll go ahead and fix that and fix the conflicts with main prior to merging.

**Review by hipsterusername — APPROVED** [review]
